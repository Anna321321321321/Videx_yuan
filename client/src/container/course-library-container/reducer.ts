import immutable, { fromJS } from 'immutable';
import * as actions from './actions';

const INITIAL_STATE: immutable.Map<any, any> = fromJS({
  courses: [],
  metadata: {
    adminAccess: false
  },
  initialized: false
});

export const reducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case actions.types.INIT:
      return state.merge(action.payload).set('initialized', true);

    case actions.types.DEINIT:
      return INITIAL_STATE;

    default:
      return state;
  }
};
