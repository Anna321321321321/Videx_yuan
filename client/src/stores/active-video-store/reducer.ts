import immutable, { fromJS } from 'immutable';
import * as actionTypes from './actionTypes';
import _ from 'lodash';

const INITIAL_STATE: immutable.Map<any, any> = fromJS({
  courseId: null,
  id: null,
  name: null,
  duration: null,
  category: null,
  transcript: {
    text: null,
    file: null
  },
  video: {
    streaming: null
  },
  thumbnail: {
    url: null,
    height: null,
    width: null,
    sas: null
  },
  isInitialized: false,
  playlists: []
});

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case actionTypes.INIT:
      return state.merge(action.payload).set('isInitialized', true);

    case actionTypes.DEINIT:
      return INITIAL_STATE;

    case actionTypes.SET_DURATION:
      return state.set('duration', action.payload.duration);

    case actionTypes.SET_PLAYLISTS:
      return state.set('playlists', action.payload.playlist);

    default:
      return state;
  }
};
