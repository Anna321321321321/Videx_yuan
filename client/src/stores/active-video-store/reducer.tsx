import immutable, { fromJS } from 'immutable';
import React from 'react';
import { Modal } from 'antd';
import CopyText from '../../components/copy-text';
import * as actionTypes from './actionTypes';
import _ from 'lodash';

const INITIAL_STATE: immutable.Map<any, any> = fromJS({
  courseId: null,
  id: null,
  name: null,
  duration: null,
  category: null,
  transcript: {
    text: null,
    file: null,
  },
  video: {
    streaming: null,
  },
  thumbnail: {
    url: null,
    height: null,
    width: null,
    sas: null,
  },
  isInitialized: false,
  playlists: [],
  sharedAnnotations: [],
  isShared: false,
  accessedBy: [],
});

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case actionTypes.INIT:
      return state.merge(action.payload).set('isInitialized', true);

    case actionTypes.DEINIT:
      return INITIAL_STATE;

    case actionTypes.SET_DURATION:
      return state.set('duration', action.payload.duration);
    case actionTypes.SET_PLAYLISTS:
      return state.set('playlists', action.payload.playlist);
    case actionTypes.INITSHAREUPDATELINK:
      return state.set('link', action.payload.link);
    case actionTypes.SHAREPUBLIC:
      const { link } = action.payload;
      const url2 = window.location.href + '/' + link;
      Modal.info({
        title: 'You can share your public annotations.',
        content: (
          <div>
            <CopyText text={url2} />
          </div>
        ),
      });
      // prettier-ignore
      return state;
    case actionTypes.GETAFROMLINK:
      return state.set('sharedAnnotations', action.payload.annotations);
    case actionTypes.ISSHARED:
      return state.set('isShared', true);
    case actionTypes.GETACCESSOR:
      return state.set('accessedBy', action.payload);
    default:
      return state;
  }
};
