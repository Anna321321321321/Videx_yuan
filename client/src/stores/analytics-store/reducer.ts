import { fromJS } from 'immutable';
import * as ActionTypes from './actionTypes';

const INITIAL_STATE = fromJS({
  size: {
    height: null,
    width: null
  },
  initialized: false,
  analytics: {
    annotations: null,
    pauses: null,
    lessonSeeks: null
  }
});

export default function(state = INITIAL_STATE, action) {
  switch (action.type) {
    case ActionTypes.SETSIZE:
      return state
        .setIn(['size', 'width'], action.payload.width)
        .setIn(['size', 'height'], action.payload.height);

    case ActionTypes.DEINIT:
      return INITIAL_STATE;

    case ActionTypes.GET_ANALYTICS:
      let lessonCount = action.payload.lessonAnnotation.rows.reduce(
        (map, item) => {
          let obj = {
            start: 0,
            end: 0
          };
          obj['start'] = parseInt(item[0]);
          obj['end'] = parseInt(item[1]);
          map.push(obj);
          return map;
        },
        []
      );

      let lessonSeeks = action.payload.lessonSeeks.rows.reduce((map, item) => {
        let obj = {
          start: 0,
          end: 0,
          count: 0
        };
        obj['start'] = parseInt(item[0]);
        obj['end'] = parseInt(item[1]);
        obj['count'] = item[2];
        map.push(obj);
        return map;
      }, []);
      let pauseCount = action.payload.lessonPauses.rows.reduce((map, item) => {
        let obj = {
          videoTimestamp: 0,
          count: 0
        };
        obj['videoTimestamp'] = parseInt(item[0]);
        obj['count'] = parseInt(item[1]);
        map.push(obj);
        return map;
      }, []);
      return state
        .setIn(['analytics', 'annotations'], lessonCount)
        .setIn(['analytics', 'pauses'], pauseCount)
        .setIn(['analytics', 'lessonSeeks'], lessonSeeks)
        .set('isInitialized', true);

    default:
      return state;
  }
}
