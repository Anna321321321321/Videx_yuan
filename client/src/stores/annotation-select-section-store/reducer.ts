import immutable, { fromJS } from 'immutable';
import * as Logger from 'videx/client/logger';
import * as actionTypes from './actionTypes';

const INITIAL_STATE: immutable.Map<any, any> = fromJS({
  range: {
    start: null,
    end: null
  },
  source: null,
  status: 'free'
});

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case actionTypes.START:
      Logger.event('SelectSection.Start', {
        source: state.get('source'),
        from: action.payload.time
      });
      return state
        .setIn(['range', 'start'], action.payload.time)
        .setIn(['range', 'end'], action.payload.time)
        .set('source', action.payload.source)
        .set('status', 'busy');

    case actionTypes.IN_PROGRESS:
      return state.setIn(['range', 'end'], action.payload.time);

    case actionTypes.END:
      Logger.event('SelectSection.End', {
        source: state.get('source'),
        from: state.getIn(['range', 'start']),
        to: state.getIn(['range', 'end'])
      });
      if (action.payload.time > state.getIn(['range', 'start'])) {
        return state
          .setIn(['range', 'end'], action.payload.time)
          .set('status', 'end');
      } else {
        const start = state.getIn(['range', 'start']);
        return state
          .setIn(['range', 'end'], start)
          .setIn(['range', 'start'], action.payload.time)
          .set('status', 'end');
      }

    case actionTypes.CLEAR:
      return state
        .setIn(['range', 'start'], null)
        .setIn(['range', 'end'], null)
        .set('source', null)
        .set('status', 'free');

    default:
      return state;
  }
};
