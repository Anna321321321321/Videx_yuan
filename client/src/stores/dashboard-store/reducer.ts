import immutable, { fromJS } from 'immutable';
import * as actions from './actions';
import _ from 'lodash';
import moment from 'moment';
import * as helpers from './helpers';

const INITIAL_STATE: immutable.Map<any, any> = fromJS({
  subscribers: 0,
  initialized: false,
  releaseDate: null,
  lessonData: [],
  lessonCountWithNameOverDate: null,
  calendarStartDate: null,
  calendarEndDate: null,
  calendarHeatmapBlank: null,
  lessonVisitNameOverTime: [],
  lessonCountGroupByDate: null,
  lessonIdNameMap: null,
  lessonFilter: []
});

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case actions.types.INIT:
      let lessonData = new Array();
      let lessonIdNameMap = null;
      let lessonCountWithNameOverDate = null;
      let lessonCountGroupByDate = null;
      const {
        calendarHeatmapBlank,
        calendarStartDate,
        calendarEndDate
      } = helpers.getBlankDateRange();
      let subscribers = 0;
      if (
        action.payload.activeSubscribers.rows.length !== 0 &&
        action.payload.lessonVisitNameOverTime.rows.length > 0 &&
        action.payload.totalViewCountOverTime.rows.length > 0
      ) {
        subscribers = action.payload.activeSubscribers.rows[0][1];
        lessonData = action.payload.lessonData.map(lesson => {
          return {
            id: lesson.id,
            publish: lesson.publish,
            name: lesson.name
          };
        });

        lessonIdNameMap = action.payload.lessonData.reduce((map, item) => {
          map[item.id] = item.name;
          return map;
        }, {});

        // map date: count
        lessonCountGroupByDate = action.payload.totalViewCountOverTime.rows.reduce(
          (map, item) => {
            const dateKey = moment(item[0])
              .utc()
              .format('YYYY-MM-DD');
            map[dateKey] = item[1];
            return map;
          },
          {}
        );
        // merge date and lesson name
        lessonCountWithNameOverDate = helpers.getLessonVisitObjectOverTime(
          action.payload.lessonVisitNameOverTime,
          lessonCountGroupByDate,
          lessonIdNameMap
        );
      }

      const calendarHeatmapData = helpers.getHeatmapData(
        lessonCountWithNameOverDate,
        calendarHeatmapBlank
      );

      return state
        .merge(action.payload)
        .set('lessonFilter', [])
        .set('subscribers', subscribers)
        .set('calendarStartDate', calendarStartDate)
        .set('calendarEndDate', calendarEndDate)
        .set('calendarHeatmapData', calendarHeatmapData)
        .set('initialized', true)
        .set('lessonData', lessonData)
        .set('lessonCountWithNameOverDate', lessonCountWithNameOverDate)
        .set(
          'lessonVisitNameOverTime',
          action.payload.lessonVisitNameOverTime.rows.length !== 0
            ? action.payload.lessonVisitNameOverTime.rows
            : []
        )
        .set('lessonCountGroupByDate', lessonCountGroupByDate)
        .set('lessonIdNameMap', lessonIdNameMap);

    case actions.types.CALENDAR_FILTER:
      const newArray = [...action.payload];
      return state.set('lessonFilter', newArray);

    case actions.types.DEINIT:
      return INITIAL_STATE;

    default:
      return state;
  }
};
