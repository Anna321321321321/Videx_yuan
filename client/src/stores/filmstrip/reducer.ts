import { fromJS } from 'immutable';
import ActionTypes from './actionTypes';

const INITIAL_STATE = fromJS({
  size: {
    height: null,
    width: null
  },
  initialized: false
});

export default function(state = INITIAL_STATE, action) {
  switch (action.type) {
    case ActionTypes.RESIZE:
      return state
        .setIn(['size', 'height'], action.payload.height)
        .setIn(['size', 'width'], action.payload.width);

    case ActionTypes.DEINIT:
      return INITIAL_STATE;

    default:
      return state;
  }
}
