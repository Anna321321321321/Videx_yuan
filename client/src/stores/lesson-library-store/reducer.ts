import * as actionTypes from './actionTypes';
import * as helpers from './helpers';
import * as LocalStorage from '../../system/local-storage';
import _ from 'lodash';

type LessonLibraryState = {
  lessons: any[];
  metadata: object;
  selectedCategory: string;
  selectedWeek: string;
  searchedLessons: string[];
  isInitialized: boolean;
  subscribers: number;
  lessonCount: number;
  analyticsMode: boolean;
  completeWatch: number;
  showDrawer: boolean;
};

const INITIAL_STATE: LessonLibraryState = {
  lessons: [],
  metadata: {},
  selectedCategory: 'all',
  selectedWeek: null,
  searchedLessons: null,
  isInitialized: false,
  subscribers: 0,
  lessonCount: 0,
  analyticsMode: false,
  completeWatch: 0,
  showDrawer: false,
};

export default (state = INITIAL_STATE, action): LessonLibraryState => {
  switch (action.type) {
    case actionTypes.INIT:
      let activeSubscribers = 0;
      if (
        action.payload.metadata.adminAccess &&
        action.payload.activeSubscribers.rows.length !== 0
      ) {
        activeSubscribers = action.payload.activeSubscribers.rows[0][1];
        const lessonViews = helpers.arrayToHashmap(
          action.payload.sessionsPerLesson
        );

        //lessonid, unique_sessions_count
        const uniqueLessonSessionsObject = helpers.arrayToHashmap(
          action.payload.uniqueSessionsPerLesson
        );

        const pausesObject = helpers.arrayToHashmap(
          action.payload.coursePauses
        );

        // lessonId, duration
        let averageViewTime = action.payload.averageViewTime.rows.reduce(
          (map, item) => {
            const splitTime = item[1].split(':');
            if (
              parseInt(splitTime[0]) >= 0 &&
              parseInt(splitTime[0]) % 1 === 0
            ) {
              const totalTime =
                parseInt(splitTime[0]) * 3600 +
                parseInt(splitTime[1]) * 60 +
                parseInt(splitTime[2]);
              map[item[0]] = helpers.time2string(totalTime);
              return map;
            } else {
              return map;
            }
          },
          {}
        );

        action.payload.lessons.map(lesson => {
          lesson.uniqueStudents =
            uniqueLessonSessionsObject[lesson.id] !== undefined
              ? uniqueLessonSessionsObject[lesson.id]
              : 0;

          lesson['pausesCount'] = pausesObject[lesson.id] !== undefined ? pausesObject[lesson.id] : 0;

          // calculate duration in string
          if (lesson.duration !== null) {
            lesson.duration = helpers.time2string(lesson.duration);
          }
          lesson['totalViews'] =
            lessonViews[lesson.id] !== undefined ? lessonViews[lesson.id] : 0;
            lesson['averageViewTime'] = 0;
          if (_.has(averageViewTime, lesson.id)) {
            lesson['averageViewTime'] = averageViewTime[lesson.id];
          }
          return lesson;
        });
      } else {
        action.payload.lessons.map(item => {
          if (item.duration !== null) {
            item.duration = helpers.time2string(item.duration);
          }
          return item;
        });
      }

      action.payload.lessons = action.payload.lessons.map(lesson => ({
        ...lesson,
        annotations: lesson.annotations.map(annotation => ({
          ...annotation,
          transcript: helpers.mergeAnnotationTranscript(
            annotation,
            lesson.transcript
          ),
        })),
      }));

      // check if selectedCategory exists
      const selectedCategory = LocalStorage.getItem(
        LocalStorage.LocalStorageKeys.SELECTED_CATEGORY
      );

      return {
        ...state,
        subscribers: activeSubscribers,
        lessonCount: action.payload.lessons.length,
        lessons: action.payload.lessons,
        metadata: action.payload.metadata,
        isInitialized: true,
        selectedCategory: selectedCategory === null ? 'all' : selectedCategory
      };

    case actionTypes.SEARCH_SUCCESS:
      return {
        ...state,
        searchedLessons: action.payload
      };

    case actionTypes.SEARCH_RESET:
      return {
        ...state,
        searchedLessons: null
      };

    case actionTypes.SELECT_WEEK:
      return {
        ...state,
        selectedWeek: action.payload
      };

    case actionTypes.SELECT_CATEGORY:
      // change in localStorage
      LocalStorage.setItem(
        LocalStorage.LocalStorageKeys.SELECTED_CATEGORY,
        action.payload
      );
      return {
        ...state,
        selectedCategory: action.payload
      };

    case actionTypes.DEINIT:
      return INITIAL_STATE;

    case actionTypes.SHOW_ANALYTICS:
      return {
        ...state,
        analyticsMode: !state.analyticsMode
      };

    case actionTypes.SHOW_DRAWER:
      return {
        ...state,
        showDrawer: !state.showDrawer
      };

    default:
      return state;
  }
};
