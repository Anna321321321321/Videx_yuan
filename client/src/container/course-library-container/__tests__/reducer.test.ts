import { fromJS } from 'immutable';
import { reducer } from '../reducer';

const unInitializedStates = fromJS({
  courses: [],
  metadata: {
    adminAccess: false
  },
  initialized: false
});

const initializedStates = fromJS({
  courses: [
    {
      name: 'test1',
      metadata: {
        contributorAccess: true,
        ownerAccess: true
      },
      id: '1'
    }
  ],
  metadata: {
    adminAccess: true
  },
  initialized: true
});

describe('User reducer unit test', () => {
  it('INIT', () => {
    let state = unInitializedStates;
    state = reducer(state, {
      type: 'courses-library/INIT',
      payload: {
        courses: [
          {
            name: 'test1',
            metadata: {
              contributorAccess: true,
              ownerAccess: true
            },
            id: '1'
          }
        ],
        metadata: {
          adminAccess: true
        }
      }
    });
    expect(state).toEqual(initializedStates);
    expect(state).toMatchSnapshot();
  });

  it('DEINIT', () => {
    let state = initializedStates;
    state = reducer(state, {
      type: 'courses-library/DEINIT'
    });
    expect(state).toEqual(unInitializedStates);
    expect(state).toMatchSnapshot();
  });
});
