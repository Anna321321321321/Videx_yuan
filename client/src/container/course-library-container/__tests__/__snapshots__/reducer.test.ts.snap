// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User reducer unit test DEINIT 1`] = `
Immutable.Map {
  "courses": Immutable.List [],
  "metadata": Immutable.Map {
    "adminAccess": false,
  },
  "initialized": false,
}
`;

exports[`User reducer unit test INIT 1`] = `
Immutable.Map {
  "courses": Immutable.List [
    Immutable.Map {
      "name": "test1",
      "metadata": Immutable.Map {
        "contributorAccess": true,
        "ownerAccess": true,
      },
      "id": "1",
    },
  ],
  "metadata": Immutable.Map {
    "adminAccess": true,
  },
  "initialized": true,
}
`;
